### PR TITLE
Langchain labs

### PR DESCRIPTION
**Pull request by krab7191**
- Basic introduction based on the dedicated squad lab
- Summarization lab based on various and sundry docs across the interwebs

### CODE REVIEW

**Review by jbukuts — COMMENTED** [review]
Cool lab. I wasn't too familiar with langchain so it's cool to see something like map-reduce implemented in this lab.  

I will note I ran into a Python issue around Chroma DB similar to what's outlined here:

https://github.com/langchain-ai/langchain/issues/7482

They noted a fix had been merged into a newer version so I bumped my `langchain` version from `0.0.236` to `0.0.312` and had no issues after that. I was using Python version 3.11.5.

I'm not sure if this warrants an update to the `requirements.txt` file though. 

Otherwise LGTM.

**Comment by timroster** [comment]
I recently updated the langchain dependency for the internal bootcamp to `0.0.312` due to the security issues with previous versions. This tested successfully across all of those labs. I think it may be a good plan to test and bump this in `requirements.txt` .

File: content/Watsonx/WatsonxAI/105/langchain-intro.ipynb [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Initialize the WatsonX model\n",
+    "params = {\n",
+    "    GenParams.DECODING_METHOD: \"sample\",\n",
+    "    GenParams.TEMPERATURE: 0.2,\n",
+    "    GenParams.TOP_P: 1,\n",
+    "    GenParams.TOP_K: 25,\n",
+    "    # GenParams.RANDOM_SEED: 42,\n",
```
Comment (ridgill) [inline content/Watsonx/WatsonxAI/105/langchain-intro.ipynb:75]: Minor: should we remove the comment?

File: content/Watsonx/WatsonxAI/105/langchain-summarization.ipynb [hunk 2]
```diff
@@ -0,0 +1,313 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 1. Summarization with langchain\n",
+    "\n",
+    "Summarization of long documents is a common LLM use case. The issue that most often arises, however is that there is a token limit for the model. (Max context window length) With langchain this can be worked around by chunking and recursive summarization."
```
[the remark this answers is not in the document]
Comment (ridgill) [reply content/Watsonx/WatsonxAI/105/langchain-summarization.ipynb:9]: Also, should the period should come after the "(Max content window length)"?

File: content/Watsonx/WatsonxAI/105/langchain-intro.ipynb [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+   "cell_type": "markdown",
+   "id": "a52e2e8f",
+   "metadata": {},
+   "source": [
+    "## 2. Prompt Templates & Chains\n",
+    "\n",
+    "In the previous example, the user input is sent directly to the Watsonx LLM, without using Langchain. This is a basic use case, but real applications are rarely so simple. When using an LLM in an application, you will usually need to reuse the same prompt across multiple scenarios. We will now replicate the previous example, but use an LLM chain. This allows us to:\n",
+    "\n",
+    "- Accept user input and contruct a prompt\n",
+    "- Generate mutiple prompts from an collection of data points in a dataset "
```
Comment (tkmonson) [inline content/Watsonx/WatsonxAI/105/langchain-intro.ipynb:119]: "an collection" -> "a collection"

File: content/Watsonx/WatsonxAI/105/langchain-intro.ipynb [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+    "  print(e)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ed7c152d",
+   "metadata": {},
+   "source": [
+    "## 4. Easy Loading of Documents Using Lang Chain\n",
+    "LangChain makes it easy to extract passages from documents so that you can answering questions based on your document's content."
```
Comment (tkmonson) [inline content/Watsonx/WatsonxAI/105/langchain-intro.ipynb:238]: "answering" -> "answer"

File: content/Watsonx/WatsonxAI/105/langchain-summarization.ipynb [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+    "reduce_template = \"\"\"The following is set of summaries:\n",
+    "{doc_summaries}\n",
+    "Take these and distill it into a final, consolidated summary of the main themes. \n",
+    "Helpful Answer:\"\"\"\n",
+    "reduce_prompt = PromptTemplate.from_template(reduce_template)\n",
+    "print(\"Init reduce chain...\")\n",
+    "reduce_chain = LLMChain(llm=flan_model, prompt=reduce_prompt)\n",
+    "\n",
+    "# Takes a list of documents, combines them into a single string, and passes this to an LLMChain\n",
+    "print(\"Stuff docuements using reduce chain...\")\n",
```
Comment (ridgill) [inline content/Watsonx/WatsonxAI/105/langchain-summarization.ipynb:232]: "documents"

**Review by tkmonson — COMMENTED** [review]
Everything worked pretty smoothly for me. Some of the outputs of the models were not very good, but not sure that we can do much about that. I got some false answers, and the summarization output from the map-reduce document chain example was not good (included the titles of the articles, which is not helpful).

**Comment by ridgill** [comment]
Awesome lab!  The only hiccup I encountered was the very last step in the summarization notebook.  
The first time I ran it there seemed to be no output beyond this one line.  Running a second time gives a more reasonable output.  Including both screenshots.  Not sure what would account for the first output:

Output1:
Screenshot 2023-10-12 at 1 17 17 PM

Output2:
Screenshot 2023-10-12 at 1 18 51 PM

**Comment by krab7191** [comment]
- Fixed typos
- Added print out for summarization intermediate steps. Note: the chunking and tokenization seems to be the weak link here. I'll continue trying options around that to get a more consistent / better output.

**Review by ridgill — APPROVED** [review]
LGTM!

**Pull request merged**